scripts/skills/perks/perk_rf_flail_spinner.nut

```diff
@@ -14,84 +16,88 @@ this.perk_rf_flail_spinner <- ::inherit("scripts/skills/skill", {
-					{
-						this.logDebug("[" + user.getName() + "] is Spinning The Flail on target [" + targetEntity.getName() + "] with skill [" + _skill.getName() + "]");
-						if (!user.isHiddenToPlayer() && _targetTile.IsVisibleForPlayer)
-						{
-							::Tactical.EventLog.log(::Const.UI.getColorizedEntityName(user) + " is Spinning the Flail");
-						}
-
-						perk.m.IsSpinningFlail = true;
-						// Increment the vanilla SkillCounter because we want this attack to trigger effects/perks
-						::Const.SkillCounter++;
-						_skill.useForFree(_targetTile);
-						perk.m.IsSpinningFlail = false;
-					}
-
-					this.getContainer().setBusy(false);
-
-				}.bindenv(this), this);
-			}
-			else
-			{
-				if (user.isAlive() && targetEntity.isAlive())
-				{
-					this.logDebug("[" + user.getName() + "] is Spinning The Flail on target [" + targetEntity.getName() + "] with skill [" + _skill.getName() + "]");
-					this.m.IsSpinningFlail = true;
-
-					// Increment the vanilla SkillCounter because we want this attack to trigger effects/perks
-					::Const.SkillCounter++;
-					_skill.useForFree(_targetTile);
-					this.m.IsSpinningFlail = false;
-				}
-			}
-		}
+		if (_targetEntity == null || !_targetEntity.isAlive() || ::Math.rand(1, 100) > this.m.Chance)
+			return;
+
+		// Unless an ACTUAL new skill was executed, we don't want to spin the flail again
+		// i.e. only spin it ONCE per actual skill execution.
+		if (::Const.SkillCounter == this.m.__SkillCount)
+			return;
+
+		this.m.__SkillCount = ::Const.SkillCounter;
+
+		this.getContainer().setBusy(true);
+		local tag = {
+			Skill = _skill,
+			User = user,
+			TargetEntity = _targetEntity,
+			TargetTile = _targetTile,
+			Callback = this.trySpinFlail.bindenv(this)
+		};
+		// Schedule with delay of 1 when in fog of war to speed up combat
+		::Time.scheduleEvent(::TimeUnit.Virtual, !user.isHiddenToPlayer() || _targetTile.IsVisibleForPlayer ? 300 : 1, tag.Callback, tag);
 	}
 
-	function onAnySkillUsed( _skill, _targetEntity, _properties )
+	function trySpinFlail( _tag )
 	{
-		if (this.m.IsSpinningFlail)
+		if (!_tag.User.isAlive() || !_tag.TargetEntity.isAlive())
+			return;
+
+		// If the skill is already executing, we wait for it to complete that first.
+		if (_tag.Skill.RF_isExecuting())
 		{
-			_properties.DamageTotalMult *= this.m.DamageMult;
+			::Time.scheduleEvent(::TimeUnit.Virtual, 1, _tag.Callback, _tag);
+			return;
 		}
+
+		this.doSpinAttack(_tag.Skill, _tag.User, _tag.TargetEntity, _tag.TargetTile);
+		this.getContainer().setBusy(false);
 	}
 
-	function onAnySkillExecuted( _skill, _targetTile, _targetEntity, _forFree )
+	function doSpinAttack( _skill, _user, _targetEntity, _targetTile )
 	{
-		if (this.isEnabled() && _skill.isAttack() && _skill.m.IsWeaponSkill)
+		::logDebug(format("[%s] is using skill [%s] on target [%s (%i)] due to %s", _user.getName(), _skill.getName(), _targetEntity.getName(), _targetEntity.getID(), this.m.Name));
```

Comment:
This looks like a remaining test debug log.
You want to at least gate it with our Reforged debug flag